rpcs3/Emu/Cell/Modules/cellMic.cpp

```diff
@@ -3,180 +3,380 @@
 #include "Emu/Cell/PPUModule.h"
 
 #include "cellMic.h"
+#include <Emu/IdManager.h>
+#include <Emu/Cell/lv2/sys_event.h>
+#include <thread>
+
+#include "3rdparty/OpenAL/include/al.h"
+#include "3rdparty/OpenAL/include/alc.h"
+#include "3rdparty/OpenAL/include/alext.h"
 
 logs::channel cellMic("cellMic");
 
+void mic_thread::on_init(const std::shared_ptr<void>& _this)
+{
+	named_thread::on_init(_this);
+}
+
+void mic_thread::on_task()
+{
+	while (fxm::check<mic_thread>() && !Emu.IsStopped())
+	{
+		if (Emu.IsPaused())
+		{
+			std::this_thread::sleep_for(1ms); // hack
+			continue;
+		}
+		if (micOpened)
+		{
+			if (micStarted)
+			{
+				if (inputDevice == NULL)
+				{
+					inputDevice = alcCaptureOpenDevice(NULL, DspFrequency, AL_FORMAT_STEREO_FLOAT32, rawFrequency / 2);
+					alcCaptureStart(inputDevice);
+				}
+				if (eventQueueKey != 0)
+				{
+					int count = 0;
+					while (count < 1)
+					{
+						alcGetIntegerv(inputDevice, ALC_CAPTURE_SAMPLES, 1, &count);
+						std::this_thread::sleep_for(10ms);
+					}
+					ALuint frameSize = 2 * bitResolution / 8;
+					ALbyte *data = (ALbyte *) calloc(frameSize, count);
+					free(buffer);
+					buffer = data;
+					bufferSize = count;
+					alcCaptureSamples(inputDevice, buffer, count);
+					auto micQueue = lv2_event_queue::find(eventQueueKey);
+					micQueue->send(0, CELL_MIC_DATA, 0, 0);
+				}
+			}
+		}
+	}
+
+}
+
+/// Initialization/Shutdown Functions
+
 s32 cellMicInit()
 {
-	cellMic.warning("cellMicInit()");
-
+	cellMic.todo("cellMicInit()");
+	const auto micThread = fxm::make<mic_thread>();
+	if (!micThread)
+	{
+		return CELL_MIC_ERROR_ALREADY_INIT;
+	}
 	return CELL_OK;
 }
 
 s32 cellMicEnd()
 {
-	cellMic.warning("cellMicEnd()");
-
+	cellMic.todo("cellMicEnd()");
+	const auto micThread = fxm::withdraw<mic_thread>();
+	if (!micThread)
+	{
+		return CELL_MIC_ERROR_NOT_INIT;
+	}
 	return CELL_OK;
 }
 
-s32 cellMicOpen()
+/// Open/Close Microphone Functions
+
+s32 cellMicOpen(u32 deviceNumber, u32 sampleRate)
 {
-	UNIMPLEMENTED_FUNC(cellMic);
+	cellMic.todo("cellMicOpen(deviceNumber=%um sampleRate=%u)", deviceNumber, sampleRate);
+	const auto micThread = fxm::get<mic_thread>();
+	if (!micThread)
```

Comment:
missing braces

Reply:
IMO It looks a lot cleaner this way, and I'm not seeing anything in the coding guidelines about this.

Reply:
it really depends.
if the file already contains a lot of braces it should be used everywhere.

Reply:
what is noticeable though is that you tend to not use vertical space

Reply:
An example of this:
![image](https://user-images.githubusercontent.com/23019877/43976781-2511e7e6-9ce2-11e8-9fc8-9a4aac565d12.png)

Reply:
just look at cellMicGetFormatRaw. it doesn't have a single empty line

Reply:
not that it really matters codewise though